{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Script: Dataflow Basics\n",
    "\n",
    "Description: Notebook where we will see the functioning of each transformation discussed during the theory.\n",
    "\n",
    "EDEM. Master Data Analytics<br>\n",
    "Professor: Javi Briones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import Python Libraries\n",
    "import logging\n",
    "import apache_beam as beam\n",
    "from apache_beam.runners.interactive.interactive_runner import InteractiveRunner\n",
    "import apache_beam.runners.interactive.interactive_beam as ib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Apache Beam Basics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"../00_DocAux/.images/beam_pipeline.png\" width=\"1000\"/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Apache Beam is a unified programming model for parallel data processing, providing a set of transformations that enable efficient manipulation and processing of data."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pipeline, PCollection & PTransform"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **PCollection**: In Apache Beam, *PCollection* represents an immutable collection of data flowing through a processing pipeline.\n",
    "\n",
    "- **Pipeline**: A *Pipeline* in Apache Beam defines a data processing flow, specifying the sequence of transformations to be applied to the PCollections.\n",
    "\n",
    "- **PTransform**: *PTransform* is an abstraction in Apache Beam that encapsulates a data transformation operation. It defines how an input PCollection is transformed into an output PCollection within the pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:apache_beam.runners.interactive.interactive_environment:Dependencies required for Interactive Beam PCollection visualization are not available, please use: `pip install apache-beam[interactive]` to install necessary dependencies to enable all data visualization features.\n"
     ]
    },
    {
     "data": {
      "application/javascript": "\n        if (typeof window.interactive_beam_jquery == 'undefined') {\n          var jqueryScript = document.createElement('script');\n          jqueryScript.src = 'https://code.jquery.com/jquery-3.4.1.slim.min.js';\n          jqueryScript.type = 'text/javascript';\n          jqueryScript.onload = function() {\n            var datatableScript = document.createElement('script');\n            datatableScript.src = 'https://cdn.datatables.net/1.10.20/js/jquery.dataTables.min.js';\n            datatableScript.type = 'text/javascript';\n            datatableScript.onload = function() {\n              window.interactive_beam_jquery = jQuery.noConflict(true);\n              window.interactive_beam_jquery(document).ready(function($){\n                \n              });\n            }\n            document.head.appendChild(datatableScript);\n          };\n          document.head.appendChild(jqueryScript);\n        } else {\n          window.interactive_beam_jquery(document).ready(function($){\n            \n          });\n        }"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('En', 1)\n",
      "('un', 2)\n",
      "('lugar', 1)\n",
      "('de', 12)\n",
      "('la', 1)\n",
      "('Mancha,', 1)\n",
      "('cuyo', 1)\n",
      "('nombre', 1)\n",
      "('no', 2)\n",
      "('quiero', 1)\n",
      "('acordarme,', 1)\n",
      "('ha', 1)\n",
      "('mucho', 1)\n",
      "('tiempo', 1)\n",
      "('que', 2)\n",
      "('vivía', 1)\n",
      "('hidalgo', 1)\n",
      "('los', 5)\n",
      "('lanza', 1)\n",
      "('en', 1)\n",
      "('astillero,', 1)\n",
      "('adarga', 1)\n",
      "('antigua,', 1)\n",
      "('rocín', 1)\n",
      "('flaco', 1)\n",
      "('y', 2)\n",
      "('galgo', 1)\n",
      "('corredor.', 1)\n",
      "('Una', 1)\n",
      "('olla', 1)\n",
      "('algo', 1)\n",
      "('más', 3)\n",
      "('vaca', 1)\n",
      "('carnero,', 1)\n",
      "('salpicón', 1)\n",
      "('las', 3)\n",
      "('noches,', 1)\n",
      "('duelos', 1)\n",
      "('quebrantos', 1)\n",
      "('sábados,', 1)\n",
      "('lantejas', 1)\n",
      "('viernes,', 1)\n",
      "('algún', 1)\n",
      "('palomino', 1)\n",
      "('añadidura', 1)\n",
      "('domingos,', 1)\n",
      "('consumían', 1)\n",
      "('tres', 1)\n",
      "('partes', 1)\n",
      "('su', 2)\n",
      "('hacienda.', 1)\n",
      "('El', 1)\n",
      "('resto', 1)\n",
      "('della', 1)\n",
      "('concluían', 1)\n",
      "('sayo', 1)\n",
      "('velarte,', 1)\n",
      "('calzas', 1)\n",
      "('velludo', 1)\n",
      "('para', 1)\n",
      "('fiestas', 1)\n",
      "('con', 2)\n",
      "('sus', 1)\n",
      "('pantuflos', 1)\n",
      "('lo', 2)\n",
      "('mismo,', 1)\n",
      "('días', 1)\n",
      "('entre', 1)\n",
      "('semana', 1)\n",
      "('se', 1)\n",
      "('honraba', 1)\n",
      "('vellorí', 1)\n",
      "('fino.', 1)\n"
     ]
    }
   ],
   "source": [
    "with beam.Pipeline(InteractiveRunner()) as p:\n",
    "\n",
    "    (p   \n",
    "        | \"Read Text from a File\" >> beam.io.ReadFromText('../00_DocAux/input_text.txt') # Reads the txt file - Quijote\n",
    "        | \"FlatMap\" >> beam.FlatMap(lambda z: z.split()) # Splits word after word by space\n",
    "        | \"Map\" >> beam.Map(lambda x: (x,1)) # For each entry above (each word), the word will be the key and the number 1 the value\n",
    "        | \"Combine\" >> beam.CombinePerKey(sum) # CombinePerKey sums the values of keys (equal keys)\n",
    "        | \"Print\" >> beam.Map(print))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ParDo vs Map"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "ParDo: Applies a function to each element in a data bundle, allowing for more complex and flexible operations than the Map transformation.\n",
    "\n",
    "DoFn: Defines a function that can be used in ParDo transformations to perform more advanced and customized operations on the elements of a dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8\n",
      "16\n",
      "24\n",
      "32\n",
      "40\n"
     ]
    }
   ],
   "source": [
    "# Map\n",
    "def edem_map(element, num):\n",
    "    return element * num\n",
    "\n",
    "# DoFn\n",
    "class edemDoFn(beam.DoFn):\n",
    "\n",
    "    def __init__(self, num):\n",
    "        self.num_ = num\n",
    "\n",
    "    def process(self, element):\n",
    "        yield element * self.num_\n",
    "\n",
    "# Pipeline\n",
    "with beam.Pipeline(InteractiveRunner()) as p:\n",
    "  data = (\n",
    "      p \n",
    "        | \"Create a PCollection\" >> beam.Create([1,2,3,4,5])\n",
    "        | \"Map\" >> beam.Map(edem_map, num=2)\n",
    "        | \"DoFn\" >> beam.ParDo(edemDoFn(4))\n",
    "        | \"Print\" >> beam.Map(print)\n",
    "  )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### DoFn Life Cycle\n",
    "\n",
    "The life cycle of a **DoFn** in Apache Beam refers to the phases that an instance of the DoFn class goes through from its initialization to its completion within the context of a *ParDo* transformation. Below, I describe the main stages of the DoFn life cycle:\n",
    "\n",
    "1. **Initialization: Setup**:\n",
    "\n",
    "   - Goal: This phase occurs once for each instance of DoFn before the transformation is executed.\n",
    "   - Process: Initialization tasks, such as configuring resources and preparing data that will be used during execution, are performed here.\n",
    "\n",
    "2. **Processing Elements (Process)**:\n",
    "\n",
    "   - Goal: This phase is executed for each input element in out data bundle.\n",
    "   - Process: The core processing logic is implemented in the *process(self,element)* method of the DoFn class. This method is called for each element and defines how each input is processed.\n",
    "\n",
    "3. **Start bundle or finish buncle**:\n",
    "\n",
    "   - Goal: This phase occurs once before/after all elements in a bundle (a portion of data processed in parallel) are being or have been processed.\n",
    "   - Process: Clean-up tasks and resource finalization used during bundle processing are carried out here.\n",
    "\n",
    "4. **Teardown**: \n",
    "\n",
    "   - Goal: This phase occurs once after all elements in the dataset have been processed.\n",
    "   - Process: Final cleaning and resource release tasks are executed, ensuring that all closing operations are completed successfully.\n",
    "\n",
    "The life cycle of a **DoFn** provides structured control over the execution of processing logic in *ParDo* transformations. **Each instance of DoFn is created, initialized, processes logic for each element, and is closed in a controlled manner**. This approach ensures proper resource management and facilitates the implementation of custom and clean data processing operations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructor started at: 2024-01-30 16:45:22.022405\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function annotate_downstream_side_inputs at 0x16a96c400> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function fix_side_input_pcoll_coders at 0x16a96c540> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function pack_combiners at 0x16a96cae0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function lift_combiners at 0x16a96cb80> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_sdf at 0x16a96cd60> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_gbk at 0x16a96ce00> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sink_flattens at 0x16a96cf40> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function greedily_fuse at 0x16a96cfe0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function read_to_impulse at 0x16a96d080> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function impulse_to_input at 0x16a96d120> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sort_stages at 0x16a96d3a0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function add_impulse_to_dangling_transforms at 0x16a96d4e0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function setup_timer_mapping at 0x16a96d300> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function populate_data_channel_coders at 0x16a96d440> ====================\n",
      "INFO:apache_beam.runners.worker.statecache:Creating state cache with size 104857600\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.worker_handlers:Created Worker handler <apache_beam.runners.portability.fn_api_runner.worker_handlers.EmbeddedWorkerHandler object at 0x17ae83cd0> for environment ref_Environment_default_environment_1 (beam:env:embedded_python:v1, b'')\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "worker started at: 2024-01-30 16:45:22.378612\n",
      "bundle started at: 2024-01-30 16:45:22.379851\n",
      "Processing element: EN\n",
      "Processing element: UN\n",
      "Processing element: LUGAR\n",
      "Processing element: DE\n",
      "Processing element: LA\n",
      "Processing element: MANCHA,\n",
      "Processing element: DE\n",
      "Processing element: CUYO\n",
      "Processing element: NOMBRE\n",
      "Processing element: NO\n",
      "Processing element: QUIERO\n",
      "Processing element: ACORDARME,\n",
      "Processing element: NO\n",
      "Processing element: HA\n",
      "Processing element: MUCHO\n",
      "Processing element: TIEMPO\n",
      "Processing element: QUE\n",
      "Processing element: VIVÍA\n",
      "Processing element: UN\n",
      "Processing element: HIDALGO\n",
      "Processing element: DE\n",
      "Processing element: LOS\n",
      "Processing element: DE\n",
      "Processing element: LANZA\n",
      "Processing element: EN\n",
      "Processing element: ASTILLERO,\n",
      "Processing element: ADARGA\n",
      "Processing element: ANTIGUA,\n",
      "Processing element: ROCÍN\n",
      "Processing element: FLACO\n",
      "Processing element: Y\n",
      "Processing element: GALGO\n",
      "Processing element: CORREDOR.\n",
      "Processing element: UNA\n",
      "Processing element: OLLA\n",
      "Processing element: DE\n",
      "Processing element: ALGO\n",
      "Processing element: MÁS\n",
      "Processing element: VACA\n",
      "Processing element: QUE\n",
      "Processing element: CARNERO,\n",
      "Processing element: SALPICÓN\n",
      "Processing element: LAS\n",
      "Processing element: MÁS\n",
      "Processing element: NOCHES,\n",
      "Processing element: DUELOS\n",
      "Processing element: Y\n",
      "Processing element: QUEBRANTOS\n",
      "Processing element: LOS\n",
      "Processing element: SÁBADOS,\n",
      "Processing element: LANTEJAS\n",
      "Processing element: LOS\n",
      "Processing element: VIERNES,\n",
      "Processing element: ALGÚN\n",
      "Processing element: PALOMINO\n",
      "Processing element: DE\n",
      "Processing element: AÑADIDURA\n",
      "Processing element: LOS\n",
      "Processing element: DOMINGOS,\n",
      "Processing element: CONSUMÍAN\n",
      "Processing element: LAS\n",
      "Processing element: TRES\n",
      "Processing element: PARTES\n",
      "Processing element: DE\n",
      "Processing element: SU\n",
      "Processing element: HACIENDA.\n",
      "Processing element: EL\n",
      "Processing element: RESTO\n",
      "Processing element: DELLA\n",
      "Processing element: CONCLUÍAN\n",
      "Processing element: SAYO\n",
      "Processing element: DE\n",
      "Processing element: VELARTE,\n",
      "Processing element: CALZAS\n",
      "Processing element: DE\n",
      "Processing element: VELLUDO\n",
      "Processing element: PARA\n",
      "Processing element: LAS\n",
      "Processing element: FIESTAS\n",
      "Processing element: CON\n",
      "Processing element: SUS\n",
      "Processing element: PANTUFLOS\n",
      "Processing element: DE\n",
      "Processing element: LO\n",
      "Processing element: MISMO,\n",
      "Processing element: LOS\n",
      "Processing element: DÍAS\n",
      "Processing element: DE\n",
      "Processing element: ENTRE\n",
      "Processing element: SEMANA\n",
      "Processing element: SE\n",
      "Processing element: HONRABA\n",
      "Processing element: CON\n",
      "Processing element: SU\n",
      "Processing element: VELLORÍ\n",
      "Processing element: DE\n",
      "Processing element: LO\n",
      "Processing element: MÁS\n",
      "Processing element: FINO.\n",
      "bundle finished at: 2024-01-30 16:45:22.384908\n",
      "worker finished at: 2024-01-30 16:45:22.399294\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x178bcf290>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T16:51:26.579383+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17a34c610>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T16:51:27.054592+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x177fcf550>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T16:51:31.072966+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17ae98bd0>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T16:51:34.097967+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x178bbb410>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T16:51:40.589306+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x178e07610>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:00:45.01033+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17a304610>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:01:09.16266+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17ac7eb50>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:01:14.964661+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17a020e90>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {created_time:\"2024-01-30T17:01:52.327732+01:00\", grpc_status:4, grpc_message:\"Deadline Exceeded\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17a6e6a10>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:01:52.327684+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17a304610>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:02:59.834778+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17a33bd90>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {created_time:\"2024-01-30T17:03:01.306575+01:00\", grpc_status:4, grpc_message:\"Deadline Exceeded\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17ac7eb50>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {created_time:\"2024-01-30T17:03:05.621676+01:00\", grpc_status:4, grpc_message:\"Deadline Exceeded\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17a6e6a10>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:03:36.606306+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17a020e90>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:03:36.606364+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x178f1a8d0>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {created_time:\"2024-01-30T17:08:42.354364+01:00\", grpc_status:4, grpc_message:\"Deadline Exceeded\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x178131090>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {created_time:\"2024-01-30T17:08:43.198432+01:00\", grpc_status:4, grpc_message:\"Deadline Exceeded\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17ae30f50>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:08:43.198483+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from datetime import datetime\n",
    "class DoFnLifeCycle(beam.DoFn):\n",
    "\n",
    "  def now(self):\n",
    "    self._now = datetime.now()\n",
    "    return self._now\n",
    "\n",
    "  def __init__(self):\n",
    "    print(\"Constructor started at: %s\" % self.now())\n",
    "\n",
    "  def setup(self):\n",
    "    print(\"worker started at: %s\" % self.now())\n",
    "\n",
    "  def start_bundle(self):\n",
    "    print(\"bundle started at: %s\" % self.now())\n",
    "\n",
    "  def process(self, element):\n",
    "    words = element.split()\n",
    "    for word in words:\n",
    "      print(\"Processing element: %s\" % word.upper())\n",
    "      yield word.upper()\n",
    "\n",
    "  def finish_bundle(self):\n",
    "    print(\"bundle finished at: %s\" % self.now())\n",
    "\n",
    "  def teardown(self):\n",
    "    print(\"worker finished at: %s\" % self.now())\n",
    "\n",
    "with beam.Pipeline(InteractiveRunner()) as p:\n",
    "  input_data = (\n",
    "      p \n",
    "        | \"Reading the input file\" >> beam.io.ReadFromText('../00_DocAux/input_text.txt')\n",
    "        | \"DoFn Life Cycle\" >> beam.ParDo(DoFnLifeCycle())\n",
    "  )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Transformations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- GroupByKey\n",
    "\n",
    "It Groups the elements of a data bundle according to a common key, generating a set where the keys are unique, and the values are lists of elements associated with each key."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function annotate_downstream_side_inputs at 0x16a96c400> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function fix_side_input_pcoll_coders at 0x16a96c540> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function pack_combiners at 0x16a96cae0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function lift_combiners at 0x16a96cb80> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_sdf at 0x16a96cd60> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_gbk at 0x16a96ce00> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sink_flattens at 0x16a96cf40> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function greedily_fuse at 0x16a96cfe0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function read_to_impulse at 0x16a96d080> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function impulse_to_input at 0x16a96d120> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sort_stages at 0x16a96d3a0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function add_impulse_to_dangling_transforms at 0x16a96d4e0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function setup_timer_mapping at 0x16a96d300> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function populate_data_channel_coders at 0x16a96d440> ====================\n",
      "INFO:apache_beam.runners.worker.statecache:Creating state cache with size 104857600\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.worker_handlers:Created Worker handler <apache_beam.runners.portability.fn_api_runner.worker_handlers.EmbeddedWorkerHandler object at 0x17afc0bd0> for environment ref_Environment_default_environment_1 (beam:env:embedded_python:v1, b'')\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Spain', ['Valencia', 'Barcelona'])\n",
      "('France', ['Paris'])\n"
     ]
    }
   ],
   "source": [
    "with beam.Pipeline(InteractiveRunner()) as p:\n",
    "\n",
    "    data = (p | \"PCollection\" >> beam.Create([('Spain', 'Valencia'), ('Spain','Barcelona'), ('France', 'Paris')]))\n",
    "\n",
    "    (data \n",
    "        | \"Combined\" >> beam.GroupByKey()\n",
    "        | \"Print\" >> beam.Map(print))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- CoGroupByKey\n",
    "\n",
    "It merges two PCollections by key, generating pairs of keys and lists of associated elements from both data bundles. Used to perform operations involving data from two different sources."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function annotate_downstream_side_inputs at 0x16a96c400> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function fix_side_input_pcoll_coders at 0x16a96c540> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function pack_combiners at 0x16a96cae0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function lift_combiners at 0x16a96cb80> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_sdf at 0x16a96cd60> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_gbk at 0x16a96ce00> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sink_flattens at 0x16a96cf40> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function greedily_fuse at 0x16a96cfe0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function read_to_impulse at 0x16a96d080> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function impulse_to_input at 0x16a96d120> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sort_stages at 0x16a96d3a0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function add_impulse_to_dangling_transforms at 0x16a96d4e0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function setup_timer_mapping at 0x16a96d300> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function populate_data_channel_coders at 0x16a96d440> ====================\n",
      "INFO:apache_beam.runners.worker.statecache:Creating state cache with size 104857600\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.worker_handlers:Created Worker handler <apache_beam.runners.portability.fn_api_runner.worker_handlers.EmbeddedWorkerHandler object at 0x17f310bd0> for environment ref_Environment_default_environment_1 (beam:env:embedded_python:v1, b'')\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('Spain', (['Valencia', 'Barcelona'], ['Madrid', 'Alicante']))\n",
      "('France', (['Paris'], ['Lyon']))\n"
     ]
    }
   ],
   "source": [
    "with beam.Pipeline(InteractiveRunner()) as p:\n",
    "\n",
    "    p1 = p | \"PCollection 01\" >> beam.Create([('Spain', 'Valencia'), ('Spain','Barcelona'), ('France', 'Paris')])\n",
    "    p2 = p | \"PCollection 02\" >> beam.Create([('Spain', 'Madrid'), ('Spain','Alicante'), ('France', 'Lyon')])\n",
    "\n",
    "    data = ((p1,p2) | beam.CoGroupByKey())\n",
    "\n",
    "    data | \"Print\" >> beam.Map(print)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Flatten\n",
    "\n",
    "It combines multiple PCollections into a single collection, flattening the nested structure."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function annotate_downstream_side_inputs at 0x16a96c400> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function fix_side_input_pcoll_coders at 0x16a96c540> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function pack_combiners at 0x16a96cae0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function lift_combiners at 0x16a96cb80> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_sdf at 0x16a96cd60> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_gbk at 0x16a96ce00> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sink_flattens at 0x16a96cf40> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function greedily_fuse at 0x16a96cfe0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function read_to_impulse at 0x16a96d080> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function impulse_to_input at 0x16a96d120> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sort_stages at 0x16a96d3a0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function add_impulse_to_dangling_transforms at 0x16a96d4e0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function setup_timer_mapping at 0x16a96d300> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function populate_data_channel_coders at 0x16a96d440> ====================\n",
      "INFO:apache_beam.runners.worker.statecache:Creating state cache with size 104857600\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.worker_handlers:Created Worker handler <apache_beam.runners.portability.fn_api_runner.worker_handlers.EmbeddedWorkerHandler object at 0x17ea4ba10> for environment ref_Environment_default_environment_1 (beam:env:embedded_python:v1, b'')\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.01 seconds.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "London\n",
      "Manchester\n",
      "Liverpool\n",
      "New York\n",
      "Los Angeles\n",
      "Miami\n",
      "Chicago\n",
      "Madrid\n",
      "Barcelona\n",
      "Valencia\n",
      "Malaga\n"
     ]
    }
   ],
   "source": [
    "with beam.Pipeline(InteractiveRunner()) as p:\n",
    "\n",
    "    p1 = p | \"PCollection 01\" >> beam.Create(['New York', 'Los Angeles', 'Miami', 'Chicago'])\n",
    "    p2 = p | \"Pcollection 02\" >> beam.Create(['Madrid', 'Barcelona', 'Valencia', 'Malaga'])\n",
    "    p3 = p | \"Pcollection 03\" >> beam.Create(['London','Manchester', 'Liverpool'])\n",
    "\n",
    "    merged = ((p1,p2,p3)| beam.Flatten())\n",
    "\n",
    "    merged | beam.Map(print)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Partition\n",
    "\n",
    "it splits a PCollection into different partitions based on certain criteria, enabling parallel and distributed processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function annotate_downstream_side_inputs at 0x16a96c400> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function fix_side_input_pcoll_coders at 0x16a96c540> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function pack_combiners at 0x16a96cae0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function lift_combiners at 0x16a96cb80> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_sdf at 0x16a96cd60> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_gbk at 0x16a96ce00> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sink_flattens at 0x16a96cf40> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function greedily_fuse at 0x16a96cfe0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function read_to_impulse at 0x16a96d080> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function impulse_to_input at 0x16a96d120> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sort_stages at 0x16a96d3a0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function add_impulse_to_dangling_transforms at 0x16a96d4e0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function setup_timer_mapping at 0x16a96d300> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function populate_data_channel_coders at 0x16a96d440> ====================\n",
      "INFO:apache_beam.runners.worker.statecache:Creating state cache with size 104857600\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.worker_handlers:Created Worker handler <apache_beam.runners.portability.fn_api_runner.worker_handlers.EmbeddedWorkerHandler object at 0x17e3ca290> for environment ref_Environment_default_environment_1 (beam:env:embedded_python:v1, b'')\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'country': 'Switzerland', 'city': 'Zurich'}\n",
      "{'country': 'Switzerland', 'city': 'Geneva'}\n"
     ]
    }
   ],
   "source": [
    "countries = ['Spain', 'USA', 'Switzerland']\n",
    "\n",
    "def partition_fn(country,num_countries):\n",
    "    return countries.index(country['country'])\n",
    "\n",
    "with beam.Pipeline(InteractiveRunner()) as p:\n",
    "\n",
    "        p1,p2,p3 = (\n",
    "                p \n",
    "                | \"PCollection\" >> beam.Create([\n",
    "                        {'country': 'Spain', 'city': 'Valencia'},\n",
    "                        {'country': 'Spain', 'city': 'Barcelona'},\n",
    "                        {'country': 'USA', 'city': 'New York'},\n",
    "                        {'country': 'Switzerland', 'city': 'Zurich'},\n",
    "                        {'country': 'Switzerland', 'city': 'Geneva'}  \n",
    "                ])\n",
    "                | \"partition\" >> beam.Partition(partition_fn, len(countries))\n",
    "        )\n",
    "\n",
    "        p3 | \"PCollection for Switzerland\" >> beam.Map(print)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Combine\n",
    "\n",
    "It combines values associated with the same key using a specific combining function, useful for performing key-based aggregate operations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function annotate_downstream_side_inputs at 0x16a96c400> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function fix_side_input_pcoll_coders at 0x16a96c540> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function pack_combiners at 0x16a96cae0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function lift_combiners at 0x16a96cb80> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_sdf at 0x16a96cd60> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function expand_gbk at 0x16a96ce00> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sink_flattens at 0x16a96cf40> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function greedily_fuse at 0x16a96cfe0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function read_to_impulse at 0x16a96d080> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function impulse_to_input at 0x16a96d120> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function sort_stages at 0x16a96d3a0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function add_impulse_to_dangling_transforms at 0x16a96d4e0> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function setup_timer_mapping at 0x16a96d300> ====================\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.translations:==================== <function populate_data_channel_coders at 0x16a96d440> ====================\n",
      "INFO:apache_beam.runners.worker.statecache:Creating state cache with size 104857600\n",
      "INFO:apache_beam.runners.portability.fn_api_runner.worker_handlers:Created Worker handler <apache_beam.runners.portability.fn_api_runner.worker_handlers.EmbeddedWorkerHandler object at 0x17f082f50> for environment ref_Environment_default_environment_1 (beam:env:embedded_python:v1, b'')\n",
      "INFO:apache_beam.io.filebasedsink:Starting finalize_write threads with num_shards: 1 (skipped: 0), batches: 1, num_threads: 1\n",
      "INFO:apache_beam.io.filebasedsink:Renamed 1 shards in 0.00 seconds.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('User1', 8)\n",
      "('User2', 5)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17ef73c50>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:20:45.560053+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17ef33450>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:20:46.569952+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17f0f7850>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {grpc_message:\"Deadline Exceeded\", grpc_status:4, created_time:\"2024-01-30T17:35:37.652637+01:00\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17f177910>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {created_time:\"2024-01-30T17:35:37.652594+01:00\", grpc_status:4, grpc_message:\"Deadline Exceeded\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17ef64050>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {created_time:\"2024-01-30T17:35:39.675689+01:00\", grpc_status:4, grpc_message:\"Deadline Exceeded\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17f01c990>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {created_time:\"2024-01-30T17:35:41.104085+01:00\", grpc_status:4, grpc_message:\"Deadline Exceeded\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n",
      "ERROR:apache_beam.runners.direct.executor:Exception at bundle <apache_beam.runners.direct.bundle_factory._Bundle object at 0x17f06de90>, due to an exception.\n",
      " Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 65, in error_remapped_callable\n",
      "    return callable_(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1160, in __call__\n",
      "    return _end_unary_response_blocking(state, call, False, None)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/grpc/_channel.py\", line 1003, in _end_unary_response_blocking\n",
      "    raise _InactiveRpcError(state)  # pytype: disable=not-instantiable\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "grpc._channel._InactiveRpcError: <_InactiveRpcError of RPC that terminated with:\n",
      "\tstatus = StatusCode.DEADLINE_EXCEEDED\n",
      "\tdetails = \"Deadline Exceeded\"\n",
      "\tdebug_error_string = \"UNKNOWN:Error received from peer  {created_time:\"2024-01-30T17:35:43.155165+01:00\", grpc_status:4, grpc_message:\"Deadline Exceeded\"}\"\n",
      ">\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 370, in call\n",
      "    self.attempt_call(\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py\", line 416, in attempt_call\n",
      "    result = evaluator.finish_bundle()\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 704, in finish_bundle\n",
      "    data = self._read_from_pubsub(self.source.timestamp_attribute)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/transform_evaluator.py\", line 691, in _read_from_pubsub\n",
      "    response = sub_client.pull(\n",
      "               ^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/pubsub_v1/services/subscriber/client.py\", line 1495, in pull\n",
      "    response = rpc(\n",
      "               ^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/gapic_v1/method.py\", line 113, in __call__\n",
      "    return wrapped_func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 349, in retry_wrapped_func\n",
      "    return retry_target(\n",
      "           ^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/retry.py\", line 191, in retry_target\n",
      "    return target()\n",
      "           ^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/timeout.py\", line 120, in func_with_timeout\n",
      "    return func(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/Users/jumepre/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/google/api_core/grpc_helpers.py\", line 67, in error_remapped_callable\n",
      "    raise exceptions.from_grpc_error(exc) from exc\n",
      "google.api_core.exceptions.DeadlineExceeded: 504 Deadline Exceeded\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with beam.Pipeline(InteractiveRunner()) as p:\n",
    "\n",
    "    data = (p | \"PCollection\" >> beam.Create([('User1', 1), ('User2', 5), ('User1', 7)]))\n",
    "\n",
    "    (data \n",
    "        | \"Combined\" >> beam.CombinePerKey(sum)\n",
    "        | \"Print\" >> beam.Map(print))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Streaming"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### gcloud commnads:\n",
    "\n",
    "- PubSub Topics\n",
    "\n",
    "```\n",
    "gcloud pubsub topics create <TOPIC_NAME>\n",
    "```\n",
    "\n",
    "- PubSub Subscriptions\n",
    "\n",
    "```\n",
    "gcloud pubsub subscriptions create <SUBSCRIPTION_NAME> --topic=<TOPIC_NAME>\n",
    "```\n",
    "\n",
    "- Google Cloud Storage Bucket  \n",
    "\n",
    "```\n",
    "gcloud storage mb gs://<BUCKET_NAME> --location=<REGION_ID>\n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Variables\n",
    "project_id = \"dataflow-edem\"\n",
    "subscription_name = \"jumepe_dataflow-sub\"\n",
    "bq_dataset = \"edem_mba\"\n",
    "bq_table = \"edem\"\n",
    "bucket_name = \"dataflow-edem\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### PubSub - Dataflow - BigQuery"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Local"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:The process started\n",
      "INFO:root:Missing pipeline option (runner). Executing pipeline using the default runner: DirectRunner.\n",
      "INFO:apache_beam.runners.direct.direct_runner:Running pipeline with DirectRunner.\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[15], line 39\u001b[0m\n\u001b[1;32m     36\u001b[0m logging\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe process started\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     38\u001b[0m \u001b[38;5;66;03m# Run Process\u001b[39;00m\n\u001b[0;32m---> 39\u001b[0m \u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[15], line 18\u001b[0m, in \u001b[0;36mrun\u001b[0;34m()\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun\u001b[39m():\n\u001b[0;32m---> 18\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mwith\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mbeam\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mPipeline\u001b[49m\u001b[43m(\u001b[49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mPipelineOptions\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstreaming\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msave_main_session\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mas\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mp\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m     19\u001b[0m \u001b[43m        \u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[43m            \u001b[49m\u001b[43mp\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     21\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;241;43m|\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mReadFromPubSub\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m>>\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mbeam\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mReadFromPubSub\u001b[49m\u001b[43m(\u001b[49m\u001b[43msubscription\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mprojects/\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mproject_id\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m/subscriptions/\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43msubscription_name\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;66;43;03m# Read from PubSub, this is our source for DF\u001b[39;49;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     29\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/pipeline.py:601\u001b[0m, in \u001b[0;36mPipeline.__exit__\u001b[0;34m(self, exc_type, exc_val, exc_tb)\u001b[0m\n\u001b[1;32m    599\u001b[0m   \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m exc_type:\n\u001b[1;32m    600\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mresult \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun()\n\u001b[0;32m--> 601\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mresult\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait_until_finish\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    602\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    603\u001b[0m   \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_extra_context\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__exit__\u001b[39m(exc_type, exc_val, exc_tb)\n",
      "File \u001b[0;32m~/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/direct_runner.py:585\u001b[0m, in \u001b[0;36mDirectPipelineResult.wait_until_finish\u001b[0;34m(self, duration)\u001b[0m\n\u001b[1;32m    582\u001b[0m   \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mNotImplementedError\u001b[39;00m(\n\u001b[1;32m    583\u001b[0m       \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDirectRunner does not support duration argument.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m    584\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 585\u001b[0m   \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_executor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mawait_completion\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    586\u001b[0m   \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_state \u001b[38;5;241m=\u001b[39m PipelineState\u001b[38;5;241m.\u001b[39mDONE\n\u001b[1;32m    587\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m:  \u001b[38;5;66;03m# pylint: disable=broad-except\u001b[39;00m\n",
      "File \u001b[0;32m~/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py:432\u001b[0m, in \u001b[0;36mExecutor.await_completion\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    431\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mawait_completion\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 432\u001b[0m   \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_executor\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mawait_completion\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py:477\u001b[0m, in \u001b[0;36m_ExecutorServiceParallelExecutor.await_completion\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    476\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mawait_completion\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[0;32m--> 477\u001b[0m   update \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvisible_updates\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtake\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    478\u001b[0m   \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    479\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m update\u001b[38;5;241m.\u001b[39mexception:\n",
      "File \u001b[0;32m~/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/direct/executor.py:551\u001b[0m, in \u001b[0;36m_ExecutorServiceParallelExecutor._TypedUpdateQueue.take\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    549\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m \u001b[38;5;28;01mTrue\u001b[39;00m:\n\u001b[1;32m    550\u001b[0m   \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 551\u001b[0m     item \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_queue\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    552\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_queue\u001b[38;5;241m.\u001b[39mtask_done()\n\u001b[1;32m    553\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m item\n",
      "File \u001b[0;32m/opt/homebrew/Cellar/python@3.11/3.11.7_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/queue.py:180\u001b[0m, in \u001b[0;36mQueue.get\u001b[0;34m(self, block, timeout)\u001b[0m\n\u001b[1;32m    178\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m remaining \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0.0\u001b[39m:\n\u001b[1;32m    179\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m Empty\n\u001b[0;32m--> 180\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnot_empty\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mremaining\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    181\u001b[0m item \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_get()\n\u001b[1;32m    182\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mnot_full\u001b[38;5;241m.\u001b[39mnotify()\n",
      "File \u001b[0;32m/opt/homebrew/Cellar/python@3.11/3.11.7_1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/threading.py:331\u001b[0m, in \u001b[0;36mCondition.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    329\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    330\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[0;32m--> 331\u001b[0m         gotit \u001b[38;5;241m=\u001b[39m \u001b[43mwaiter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43macquire\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    332\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    333\u001b[0m         gotit \u001b[38;5;241m=\u001b[39m waiter\u001b[38;5;241m.\u001b[39macquire(\u001b[38;5;28;01mFalse\u001b[39;00m)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import apache_beam as beam\n",
    "import json\n",
    "import logging\n",
    "from apache_beam.options.pipeline_options import PipelineOptions\n",
    "import sys\n",
    "\n",
    "sys.argv = ['--flexrs_goal', 'SPEED_OPTIMIZED']\n",
    "\n",
    "def decode_message(msg):\n",
    "\n",
    "    output = msg.decode('utf-8')\n",
    "\n",
    "    logging.info(\"New PubSub Message: %s\", output)\n",
    "\n",
    "    return json.loads(output)\n",
    "\n",
    "def run():\n",
    "    with beam.Pipeline(options=PipelineOptions(streaming=True, save_main_session=True)) as p:\n",
    "        (\n",
    "            p \n",
    "            | \"ReadFromPubSub\" >> beam.io.ReadFromPubSub(subscription=f'projects/{project_id}/subscriptions/{subscription_name}') # Read from PubSub, this is our source for DF\n",
    "            | \"decode msg\" >> beam.Map(decode_message) # We need to decode, otherwise we'll se bits.\n",
    "            | \"Write to BigQuery\" >> beam.io.WriteToBigQuery(\n",
    "                table = f\"{project_id}:{bq_dataset}.{bq_table}\", # Required Format: PROJECT_ID:DATASET.TABLE\n",
    "                schema='nombre:STRING', # Required Format: field:TYPE\n",
    "                create_disposition=beam.io.BigQueryDisposition.CREATE_NEVER,\n",
    "                write_disposition=beam.io.BigQueryDisposition.WRITE_APPEND\n",
    "            )\n",
    "        )\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    # Set Logs\n",
    "    logging.getLogger().setLevel(logging.INFO)\n",
    "\n",
    "    logging.info(\"The process started\")\n",
    "    \n",
    "    # Run Process\n",
    "    run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Dataflow (Google Cloud)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:The process started\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:root:Using provided Python SDK container image: gcr.io/cloud-dataflow/v1beta3/beam_python3.11_sdk:2.51.0\n",
      "INFO:root:Python SDK container image set to \"gcr.io/cloud-dataflow/v1beta3/beam_python3.11_sdk:2.51.0\" for Docker environment\n",
      "INFO:apache_beam.runners.dataflow.internal.apiclient:Starting GCS upload to gs://dataflow-edem/staging/beamapp-jumepre-0130152645-336574-h7ljea99.1706628405.336652/pipeline.pb...\n",
      "INFO:oauth2client.transport:Attempting refresh to obtain initial access_token\n",
      "INFO:oauth2client.client:Refreshing access_token\n",
      "INFO:oauth2client.transport:Attempting refresh to obtain initial access_token\n",
      "INFO:oauth2client.client:Refreshing access_token\n",
      "INFO:apache_beam.runners.dataflow.internal.apiclient:Completed GCS upload to gs://dataflow-edem/staging/beamapp-jumepre-0130152645-336574-h7ljea99.1706628405.336652/pipeline.pb in 0 seconds.\n",
      "INFO:apache_beam.runners.dataflow.internal.apiclient:Create job: <Job\n",
      " clientRequestId: '20240130152645336844-4643'\n",
      " createTime: '2024-01-30T15:26:48.049398Z'\n",
      " currentStateTime: '1970-01-01T00:00:00Z'\n",
      " id: '2024-01-30_07_26_46-2430606315461399349'\n",
      " location: 'europe-southwest1'\n",
      " name: 'beamapp-jumepre-0130152645-336574-h7ljea99'\n",
      " projectId: 'dataflow-edem'\n",
      " stageStates: []\n",
      " startTime: '2024-01-30T15:26:48.049398Z'\n",
      " steps: []\n",
      " tempFiles: []\n",
      " type: TypeValueValuesEnum(JOB_TYPE_STREAMING, 2)>\n",
      "INFO:apache_beam.runners.dataflow.internal.apiclient:Created job with id: [2024-01-30_07_26_46-2430606315461399349]\n",
      "INFO:apache_beam.runners.dataflow.internal.apiclient:Submitted job: 2024-01-30_07_26_46-2430606315461399349\n",
      "INFO:apache_beam.runners.dataflow.internal.apiclient:To access the Dataflow monitoring console, please navigate to https://console.cloud.google.com/dataflow/jobs/europe-southwest1/2024-01-30_07_26_46-2430606315461399349?project=dataflow-edem\n",
      "INFO:apache_beam.runners.dataflow.dataflow_runner:Job 2024-01-30_07_26_46-2430606315461399349 is in state JOB_STATE_PENDING\n",
      "WARNING:apache_beam.runners.dataflow.dataflow_runner:2024-01-30T15:26:48.738Z: JOB_MESSAGE_WARNING: Autoscaling is enabled for Dataflow Streaming Engine. Workers will scale between 1 and 100 unless maxNumWorkers is specified.\n",
      "INFO:apache_beam.runners.dataflow.dataflow_runner:2024-01-30T15:26:54.765Z: JOB_MESSAGE_BASIC: Worker configuration: e2-standard-2 in europe-southwest1-a.\n",
      "INFO:apache_beam.runners.dataflow.dataflow_runner:2024-01-30T15:26:56.355Z: JOB_MESSAGE_BASIC: Running job using Streaming Engine\n",
      "INFO:apache_beam.runners.dataflow.dataflow_runner:2024-01-30T15:26:56.550Z: JOB_MESSAGE_BASIC: Starting 1 workers in europe-southwest1-a...\n",
      "INFO:apache_beam.runners.dataflow.dataflow_runner:Job 2024-01-30_07_26_46-2430606315461399349 is in state JOB_STATE_RUNNING\n",
      "INFO:apache_beam.runners.dataflow.dataflow_runner:2024-01-30T15:28:37.060Z: JOB_MESSAGE_BASIC: All workers have finished the startup processes and began to receive work requests.\n",
      "WARNING:apache_beam.runners.dataflow.dataflow_runner:2024-01-30T15:28:42.764Z: JOB_MESSAGE_WARNING: The worker utilization hint isn't being used by the autoscaler for the following reason: flag enabling worker utilization hint is not turned on. To use the utilization hint for autoscaling, set a value between {lower_bound_limit} and 0.9.\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[17], line 47\u001b[0m\n\u001b[1;32m     44\u001b[0m logging\u001b[38;5;241m.\u001b[39minfo(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe process started\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     46\u001b[0m \u001b[38;5;66;03m# Run Process\u001b[39;00m\n\u001b[0;32m---> 47\u001b[0m \u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[17], line 18\u001b[0m, in \u001b[0;36mrun\u001b[0;34m()\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun\u001b[39m():\n\u001b[0;32m---> 18\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43;01mwith\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mbeam\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mPipeline\u001b[49m\u001b[43m(\u001b[49m\u001b[43moptions\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mPipelineOptions\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     19\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstreaming\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;66;43;03m# save_main_session=True\u001b[39;49;00m\n\u001b[1;32m     21\u001b[0m \u001b[43m        \u001b[49m\u001b[43mproject\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mproject_id\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m     22\u001b[0m \u001b[43m        \u001b[49m\u001b[43mrunner\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mDataflowRunner\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     23\u001b[0m \u001b[43m        \u001b[49m\u001b[43mtemp_location\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgs://\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mbucket_name\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m/tmp\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     24\u001b[0m \u001b[43m        \u001b[49m\u001b[43mstaging_location\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mgs://\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mbucket_name\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m/staging\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m     25\u001b[0m \u001b[43m        \u001b[49m\u001b[43mregion\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43meurope-southwest1\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\n\u001b[1;32m     26\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43;01mas\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mp\u001b[49m\u001b[43m:\u001b[49m\n\u001b[1;32m     27\u001b[0m \u001b[43m        \u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m     28\u001b[0m \u001b[43m            \u001b[49m\u001b[43mp\u001b[49m\u001b[43m \u001b[49m\n\u001b[1;32m     29\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;241;43m|\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mReadFromPubSub\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m>>\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mbeam\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mio\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mReadFromPubSub\u001b[49m\u001b[43m(\u001b[49m\u001b[43msubscription\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mprojects/\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mproject_id\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m/subscriptions/\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43msubscription_name\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     36\u001b[0m \u001b[43m            \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     37\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/pipeline.py:601\u001b[0m, in \u001b[0;36mPipeline.__exit__\u001b[0;34m(self, exc_type, exc_val, exc_tb)\u001b[0m\n\u001b[1;32m    599\u001b[0m   \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m exc_type:\n\u001b[1;32m    600\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mresult \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun()\n\u001b[0;32m--> 601\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mresult\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait_until_finish\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    602\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    603\u001b[0m   \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_extra_context\u001b[38;5;241m.\u001b[39m\u001b[38;5;21m__exit__\u001b[39m(exc_type, exc_val, exc_tb)\n",
      "File \u001b[0;32m~/Documents/GitHub/Serverless_EDEM_2024/edem_test/lib/python3.11/site-packages/apache_beam/runners/dataflow/dataflow_runner.py:754\u001b[0m, in \u001b[0;36mDataflowPipelineResult.wait_until_finish\u001b[0;34m(self, duration)\u001b[0m\n\u001b[1;32m    752\u001b[0m thread\u001b[38;5;241m.\u001b[39mstart()\n\u001b[1;32m    753\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m thread\u001b[38;5;241m.\u001b[39mis_alive():\n\u001b[0;32m--> 754\u001b[0m   \u001b[43mtime\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msleep\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m5.0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    756\u001b[0m \u001b[38;5;66;03m# TODO: Merge the termination code in poll_for_job_completion and\u001b[39;00m\n\u001b[1;32m    757\u001b[0m \u001b[38;5;66;03m# is_in_terminal_state.\u001b[39;00m\n\u001b[1;32m    758\u001b[0m terminated \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mis_in_terminal_state()\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:apache_beam.runners.dataflow.dataflow_runner:Job 2024-01-30_07_26_46-2430606315461399349 is in state JOB_STATE_CANCELLING\n",
      "INFO:apache_beam.runners.dataflow.dataflow_runner:2024-01-30T15:33:54.192Z: JOB_MESSAGE_BASIC: Cancel request is committed for workflow job: 2024-01-30_07_26_46-2430606315461399349.\n",
      "INFO:apache_beam.runners.dataflow.dataflow_runner:2024-01-30T15:33:54.248Z: JOB_MESSAGE_BASIC: Stopping worker pool...\n",
      "INFO:apache_beam.runners.dataflow.dataflow_runner:2024-01-30T15:33:54.273Z: JOB_MESSAGE_BASIC: Stopping worker pool...\n",
      "INFO:apache_beam.runners.dataflow.dataflow_runner:2024-01-30T15:34:24.860Z: JOB_MESSAGE_BASIC: Worker pool stopped.\n"
     ]
    }
   ],
   "source": [
    "import apache_beam as beam\n",
    "import json\n",
    "import logging\n",
    "from apache_beam.options.pipeline_options import PipelineOptions\n",
    "import sys\n",
    "\n",
    "sys.argv = ['--flexrs_goal', 'SPEED_OPTIMIZED']\n",
    "\n",
    "def decode_message(msg):\n",
    "\n",
    "    output = msg.decode('utf-8')\n",
    "\n",
    "    logging.info(\"New PubSub Message: %s\", output)\n",
    "\n",
    "    return json.loads(output)\n",
    "\n",
    "def run():\n",
    "    with beam.Pipeline(options=PipelineOptions(\n",
    "        streaming=True,\n",
    "        # save_main_session=True\n",
    "        project=project_id,\n",
    "        runner=\"DataflowRunner\",\n",
    "        temp_location=f\"gs://{bucket_name}/tmp\",\n",
    "        staging_location=f\"gs://{bucket_name}/staging\",\n",
    "        region=\"europe-southwest1\"\n",
    "    )) as p:\n",
    "        (\n",
    "            p \n",
    "            | \"ReadFromPubSub\" >> beam.io.ReadFromPubSub(subscription=f'projects/{project_id}/subscriptions/{subscription_name}')\n",
    "            | \"decode msg\" >> beam.Map(decode_message)\n",
    "            | \"Write to BigQuery\" >> beam.io.WriteToBigQuery(\n",
    "                table = \"dataflow-edem:edem_mba.edem\", # Required Format: PROJECT_ID:DATASET.TABLE\n",
    "                schema='nombre:STRING', # Required Format: field:TYPE\n",
    "                create_disposition=beam.io.BigQueryDisposition.CREATE_NEVER,\n",
    "                write_disposition=beam.io.BigQueryDisposition.WRITE_APPEND\n",
    "            )\n",
    "        )\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    # Set Logs\n",
    "    logging.getLogger().setLevel(logging.INFO)\n",
    "\n",
    "    logging.info(\"The process started\")\n",
    "    \n",
    "    # Run Process\n",
    "    run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Window in Apache Beam:**\n",
    "A window in Apache Beam defines a time frame for **organizing and grouping data elements** during processing, enabling time-based specific operations.\n",
    "\n",
    "**Types of Windows:**\n",
    "\n",
    "- **Fixed Window:** Groups elements into fixed time intervals, dividing the PCollection into time-based windows.\n",
    "\n",
    "- **Sliding Window:** Allows overlapping windows, specified by a size and a stride, making it easy to analyze data in continuous intervals over time.\n",
    "\n",
    "- **Session Window:** Groups data elements that share a contiguous temporal relationship, where continuity is defined by the **inactivity gap between elements**. This dynamic window is formed based on the inactivity time between events, allowing the capture of logical sessions in data streams."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Fixed Window"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"../00_DocAux/.images/fixed_window.png\" width=\"500\"/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import apache_beam as beam\n",
    "import json\n",
    "import logging\n",
    "from apache_beam.options.pipeline_options import PipelineOptions\n",
    "\n",
    "def decode_message(msg):\n",
    "\n",
    "    output = msg.decode('utf-8')\n",
    "\n",
    "    logging.info(\"New PubSub Message: %s\", output)\n",
    "\n",
    "    return json.loads(output)['temp']\n",
    "\n",
    "class OutputDoFn(beam.DoFn):\n",
    "\n",
    "    def process(self, element):\n",
    "        yield element\n",
    "\n",
    "def run():\n",
    "    with beam.Pipeline(options=PipelineOptions(streaming=True, save_main_session=True)) as p:\n",
    "        (\n",
    "            p \n",
    "            | \"ReadFromPubSub\" >> beam.io.ReadFromPubSub(subscription=f'projects/{project_id}/subscriptions/{subscription_name}')\n",
    "            | \"Decode msg\" >> beam.Map(decode_message)\n",
    "            | \"Fixed Window\" >> beam.WindowInto(beam.window.FixedWindows(10))\n",
    "            | \"Combine\" >> beam.CombineGlobally(sum).without_defaults()\n",
    "            | \"Print\" >> beam.Map(print)\n",
    "        )\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    # Set Logs\n",
    "    logging.getLogger().setLevel(logging.INFO)\n",
    "\n",
    "    logging.info(\"The process started\")\n",
    "    \n",
    "    # Run Process\n",
    "    run()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Sliding windows "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"../00_DocAux/.images/sliding_window.png\" width=\"500\"/>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import apache_beam as beam\n",
    "import json\n",
    "import logging\n",
    "from apache_beam.options.pipeline_options import PipelineOptions\n",
    "\n",
    "def decode_message(msg):\n",
    "\n",
    "    output = msg.decode('utf-8')\n",
    "\n",
    "    logging.info(\"New PubSub Message: %s\", output)\n",
    "\n",
    "    return output['temp']\n",
    "\n",
    "class OutputDoFn(beam.DoFn):\n",
    "\n",
    "    def process(self, element):\n",
    "        yield element\n",
    "\n",
    "def run():\n",
    "    with beam.Pipeline(options=PipelineOptions(streaming=True, save_main_session=True)) as p:\n",
    "        (\n",
    "            p \n",
    "            | \"ReadFromPubSub\" >> beam.io.ReadFromPubSub(subscription=f'projects/{project_id}/subscriptions/{subscription_name}')\n",
    "            | \"Decode msg\" >> beam.Map(decode_message)\n",
    "            | \"Sliding Window\" >> beam.WindowInto(beam.window.SlidingWindows(size=60, period=20))\n",
    "            | \"Combine\" >> beam.CombinePerKey(sum)\n",
    "            | \"Print\" >> beam.Map(print)\n",
    "        )\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    # Set Logs\n",
    "    logging.getLogger().setLevel(logging.INFO)\n",
    "\n",
    "    logging.info(\"The process started\")\n",
    "    \n",
    "    # Run Process\n",
    "    run()"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "bb80d8843f3fcc2b8a9e4270eecba2ac31de31ca61bfe968e7b44b35850adf72"
  },
  "kernelspec": {
   "display_name": "Python 3.10.9 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
